src/engine/Server.cpp

```diff
@@ -833,6 +858,79 @@ Awaitable<void> Server::processQuery(
+    ad_utility::Timer& requestTimer, TimeLimit timeLimit, auto& messageSender,
+    ad_utility::SharedCancellationHandle cancellationHandle,
+    DeltaTriples& deltaTriples) {
+  auto [pinSubtrees, pinResult] = determineResultPinning(params);
+  LOG(INFO) << "Processing the following SPARQL update:"
+            << (pinResult ? " [pin result]" : "")
+            << (pinSubtrees ? " [pin subresults]" : "") << "\n"
+            << update << std::endl;
+  QueryExecutionContext qec(index_, &cache_, allocator_,
+                            sortPerformanceEstimator_, std::ref(messageSender),
+                            pinSubtrees, pinResult);
+  auto plannedQuery = co_await setupPlannedQuery(
+      params, update, qec, cancellationHandle, timeLimit, requestTimer);
+  auto qet = plannedQuery.queryExecutionTree_;
+
+  if (!plannedQuery.parsedQuery_.hasUpdateClause()) {
+    throw std::runtime_error(
+        absl::StrCat("SPARQL UPDATE was request via the HTTP request, but the "
+                     "following query was sent instead of an update: ",
+                     plannedQuery.parsedQuery_._originalString));
+  }
+  ExecuteUpdate::executeUpdate(index_, plannedQuery.parsedQuery_, qet,
+                               deltaTriples, cancellationHandle);
+
+  LOG(INFO) << "Done processing update"
+            << ", total time was " << requestTimer.msecs().count() << " ms"
+            << std::endl;
+  LOG(DEBUG) << "Runtime Info:\n"
+             << qet.getRootOperation()->runtimeInfo().toString() << std::endl;
+}
+
+// ____________________________________________________________________________
+Awaitable<void> Server::processUpdate(
+    const ad_utility::url_parser::ParamValueMap& params, const string& update,
+    ad_utility::Timer& requestTimer,
+    const ad_utility::httpUtils::HttpRequest auto& request, auto&& send,
+    TimeLimit timeLimit) {
+  auto messageSender = createMessageSender(queryHub_, request, update);
+
+  auto [cancellationHandle, cancelTimeoutOnDestruction] =
+      setupCancellationHandle(messageSender.getQueryId(), timeLimit);
+
+  // Update the delta triples.
+  co_await computeInNewThread(
+      updateThreadPool_,
+      [this, &params, &update, &requestTimer, &timeLimit, &messageSender,
+       &cancellationHandle] {
+        index_.deltaTriplesManager().modify(
+            [this, &params, &update, &requestTimer, &timeLimit, &messageSender,
+             &cancellationHandle](auto& deltaTriples) {
+              // TODO<qup42> this must also be possible in a nicer way
+              auto coro = processUpdateImpl(params, update, requestTimer,
+                                            timeLimit, messageSender,
+                                            cancellationHandle, deltaTriples);
+              boost::asio::io_context ioContext;
+              boost::asio::co_spawn(ioContext, std::move(coro),
+                                    boost::asio::detached);
+              ioContext.run();
```

Comment:
The correct way is the following:
1. Make `parseAndPlan` as well as `setupPlannedQuery` synchronous functions (no coroutines). They are currently only coroutines because of the `computeInNewThread` which is anyway undesirable for the update case.
2. In the query (not update) execution, wrap the `setupPlannedQuery` in a `computeInNewThread(queryPool)`.
3. make `processUpdateImpl` also a normal synchronous function, and simply call it here.

(I found out that the actual execution of queries currently doesn't happen on the query trees, which might lead to deadlocks... But that is another issue for another PR, but good to have looked at the code again.